File: java/src/main/java/com/authdog/resources/ProvisioningTokensResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * SCIM and HRIS provisioning token operations.
 */
public final class ProvisioningTokensResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a provisioning tokens helper.
     * @param clientParam Authdog client
     */
    public ProvisioningTokensResource(
            final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List HRIS tokens.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return tokens envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listHris(final String tenantId,
                             final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/hris-tokens",
                null, null, JsonNode.class);
    }

    /**
     * Create an HRIS token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created token envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode createHris(final String tenantId,
                               final String environmentId,
                               final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/hris-tokens",
                body, null, JsonNode.class);
    }

    /**
     * Revoke an HRIS token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param tokenId token ID
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revokeHris(final String tenantId,
                               final String environmentId,
                               final String tokenId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/hris-tokens/" + tokenId + "/revoke",
                null, null, JsonNode.class);
    }

    /**
     * Rotate an HRIS token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param tokenId token ID
     * @return rotate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode rotateHris(final String tenantId,
                               final String environmentId,
                               final String tokenId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/hris-tokens/" + tokenId + "/rotate",
                null, null, JsonNode.class);
    }

    /**
     * List SCIM tokens.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return tokens envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listScim(final String tenantId,
                             final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/scim-tokens",
                null, null, JsonNode.class);
    }

    /**
     * Create a SCIM token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return created token envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode createScim(final String tenantId,
                               final String environmentId,
                               final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/scim-tokens",
                body, null, JsonNode.class);
    }

    /**
     * Revoke a SCIM token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param tokenId token ID
     * @return revoke envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode revokeScim(final String tenantId,
                               final String environmentId,
                               final String tokenId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/scim-tokens/" + tokenId + "/revoke",
                null, null, JsonNode.class);
    }

    /**
     * Rotate a SCIM token.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param tokenId token ID
     * @return rotate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode rotateScim(final String tenantId,
                               final String environmentId,
                               final String tokenId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/scim-tokens/" + tokenId + "/rotate",
                null, null, JsonNode.class);
    }
}

File: csharp/Resources/ProvisioningTokensResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// SCIM and HRIS provisioning-token Wave 3 operations
    /// </summary>
    public class ProvisioningTokensResource
    {
        private readonly AuthdogClient _client;

        public ProvisioningTokensResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListHrisAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/hris-tokens");

        public JObject ListHris(string tenantId, string environmentId) =>
            ListHrisAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> CreateHrisAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/hris-tokens",
                body);

        public JObject CreateHris(string tenantId, string environmentId, object body) =>
            CreateHrisAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeHrisAsync(string tenantId, string environmentId, string tokenId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/hris-tokens/{tokenId}/revoke");

        public JObject RevokeHris(string tenantId, string environmentId, string tokenId) =>
            RevokeHrisAsync(tenantId, environmentId, tokenId).GetAwaiter().GetResult();

        public Task<JObject> RotateHrisAsync(string tenantId, string environmentId, string tokenId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/hris-tokens/{tokenId}/rotate");

        public JObject RotateHris(string tenantId, string environmentId, string tokenId) =>
            RotateHrisAsync(tenantId, environmentId, tokenId).GetAwaiter().GetResult();

        public Task<JObject> ListScimAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/scim-tokens");

        public JObject ListScim(string tenantId, string environmentId) =>
            ListScimAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> CreateScimAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/scim-tokens",
                body);

        public JObject CreateScim(string tenantId, string environmentId, object body) =>
            CreateScimAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> RevokeScimAsync(string tenantId, string environmentId, string tokenId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/scim-tokens/{tokenId}/revoke");

        public JObject RevokeScim(string tenantId, string environmentId, string tokenId) =>
            RevokeScimAsync(tenantId, environmentId, tokenId).GetAwaiter().GetResult();

        public Task<JObject> RotateScimAsync(string tenantId, string environmentId, string tokenId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/scim-tokens/{tokenId}/rotate");

        public JObject RotateScim(string tenantId, string environmentId, string tokenId) =>
            RotateScimAsync(tenantId, environmentId, tokenId).GetAwaiter().GetResult();
    }
}
